drivers/wifi/infineon/airoc_whd_hal_spi.c

```diff
@@ -0,0 +1,289 @@
+		.dev_gpio_sel = DEFAULT_OOB_PIN,
+		.is_falling_edge =
+			(CY_WIFI_HOST_WAKE_IRQ_EVENT == GPIO_INT_TRIG_LOW) ? WHD_TRUE : WHD_FALSE,
+		.intr_priority = CY_WIFI_OOB_INTR_PRIORITY};
+	whd_spi_config.oob_config = oob_config;
+#endif
+#if defined(SPI_DATA_IRQ_SHARED)
+	data->prev_irq_state = 0;
+#endif
+
+	/* Pull bus select line low before enabling WiFi chip */
+	gpio_pin_configure_dt(&config->bus_select_gpio, GPIO_OUTPUT_INACTIVE);
+	gpio_pin_set_dt(&config->bus_select_gpio, 0);
+
+	if (airoc_wifi_power_on(dev)) {
+		LOG_ERR("airoc_wifi_power_on returns fail");
+		return -ENODEV;
+	}
+
+	if (!spi_is_ready_dt(&config->bus_dev.bus_spi)) {
+		LOG_ERR("SPI device is not ready");
+		return -ENODEV;
+	}
+
+	/* Init wifi host driver (whd) */
+	cy_rslt_t whd_ret = whd_init(&data->whd_drv, &init_config_default,
+					&resource_ops, buffer_if, netif_funcs);
+	if (whd_ret == CY_RSLT_SUCCESS) {
+		whd_ret = whd_bus_spi_attach(data->whd_drv, &whd_spi_config,
+					    (whd_spi_t)&config->bus_dev.bus_spi);
+
+		if (whd_ret == CY_RSLT_SUCCESS) {
+			whd_ret = whd_wifi_on(data->whd_drv, interface);
+		}
+
+		if (whd_ret != CY_RSLT_SUCCESS) {
+			whd_deinit(*interface);
+			return -ENODEV;
+		}
+	}
+	return 0;
+}
+
+/*
+ * Implement SPI Transfer wrapper
+ */
+
+whd_result_t whd_bus_spi_transfer(whd_driver_t whd_driver, const uint8_t *tx, size_t tx_length,
+				  uint8_t *rx, size_t rx_length, uint8_t write_fill)
+{
+	const struct spi_dt_spec *spi_obj = whd_driver->bus_priv->spi_obj;
+	int ret;
+	whd_result_t whd_ret = WHD_SUCCESS;
+
+#if defined(SPI_DATA_IRQ_SHARED)
+	ret = whd_bus_spi_irq_enable(whd_driver, false);
+	if (ret) {
+		LOG_DBG("whd_bus_spi_irq_enable FAIL %d\n", ret);
+		whd_ret = WHD_WLAN_SDIO_ERROR;
+	}
+#endif
+
+	/* In some cases whd_bus_spi_protocol.c places the command at    */
+	/* the start of the rx buffer and passes NULL as the tx address, */
+	/* reusing the same buffer (and overwriting the tx data).        */
+	if (tx == NULL && tx_length > 0 && rx_length >= tx_length) {
+		tx = rx;
+	}
+
+	const struct spi_buf tx_buf = {.buf = (uint8_t *)tx, .len = tx_length};
+	const struct spi_buf_set tx_set = {.buffers = &tx_buf, .count = 1};
+	struct spi_buf rx_buf[2];
+	struct spi_buf_set rx_set = {.buffers = rx_buf, .count = 2};
+
+	if (rx != NULL) {
+		rx += tx_length;
+	}
+
+	if (rx_length >= tx_length) {
+		rx_length -= tx_length;
+	} else {
+		rx_length = 0;
+	}
+
+	if (spi_obj->config.operation & SPI_HALF_DUPLEX) {
+		rx_buf[0].buf = rx;
+		rx_buf[0].len = rx_length;
+		rx_set.count = 1;
+	} else {
+		/* Talking to a half-duplex device via a full-duplex SPI driver,  */
+		/* it's necessary to ignore the data returned during send.  The   */
+		/* SPI driver should not copy out the data if the buffer is NULL. */
+		rx_buf[0].buf = NULL;
+		rx_buf[0].len = tx_length;
+		rx_buf[1].buf = rx;
+		rx_buf[1].len = rx_length;
+	}
+
+	ret = spi_transceive_dt(spi_obj, &tx_set, &rx_set);
+	if (ret) {
+		LOG_DBG("spi_transceive FAIL %d\n", ret);
```

Comment:
A question (not a blocker): Is this the correct log level, given the message's content?

Reply:
Yeah, those should all be LOG_ERR.